daffodil-test/src/test/resources/org/apache/daffodil/section05/simple_types/SimpleTypes.tdml

```diff
@@ -1951,6 +1966,87 @@
     </tdml:infoset>
   </tdml:parserTestCase>
 
+  <!-- xs:dateTime with binaryCalendarRep='binarySeconds' -->
+  <tdml:parserTestCase name="dateTimeBin" root="dateTimeBin"
+    model="SimpleTypes-binary" description="Section 5 Schema types-dateTime - DFDL-5-016R"
+    roundTrip="true">
+
+    <tdml:document>
+      <tdml:documentPart type="bits">00000000 00000000 00000000 00111110</tdml:documentPart>
+    </tdml:document>
+    <tdml:infoset>
+      <tdml:dfdlInfoset>
+        <dateTimeBin>1977-01-01T00:01:09.000000+00:00</dateTimeBin>
+      </tdml:dfdlInfoset>
+    </tdml:infoset>
+  </tdml:parserTestCase>
+
+  <tdml:parserTestCase name="dateTimeBin2" root="dateTimeBin2"
+    model="SimpleTypes-binary" description="Section 5 Schema types-dateTime - DFDL-5-016R"
+    roundTrip="false">
+
+    <tdml:document>
+      <tdml:documentPart type="bits">00000000 00000000 00000000 00111101</tdml:documentPart>
+    </tdml:document>
+    <tdml:errors>
+      <tdml:error>Schema Definition Error: lengthUnits must be 'bytes' when binaryCalendarRep='binarySeconds'</tdml:error>
+    </tdml:errors>
+  </tdml:parserTestCase>
+
+  <tdml:parserTestCase name="dateTimeBin3" root="dateTimeBin3"
+    model="SimpleTypes-binary" description="Section 5 Schema types-dateTime - DFDL-5-016R"
+    roundTrip="false">
+
+    <tdml:document>
+      <tdml:documentPart type="bits">00000000 00000000 00000000 00111101</tdml:documentPart>
+    </tdml:document>
+    <tdml:errors>
+      <tdml:error>Schema Definition Error: binary xs:dateTime must be 32 bits when binaryCalendarRep='binarySeconds'</tdml:error>
+    </tdml:errors>
+  </tdml:parserTestCase>
+
+  <!-- xs:dateTime with binaryCalendarRep='binarySeconds' with implicit length -->
+  <tdml:parserTestCase name="dateTimeBin4" root="dateTimeBin4"
+    model="SimpleTypes-binary" description="Section 5 Schema types-dateTime - DFDL-5-016R"
+    roundTrip="true">
+
+    <tdml:document>
+      <tdml:documentPart type="bits">11111111 11111110 10101110 10000000</tdml:documentPart>
+    </tdml:document>
+    <tdml:infoset>
+      <tdml:dfdlInfoset>
+        <dateTimeBin4>1869-12-31T00:05:00.000000+00:00</dateTimeBin4>
+      </tdml:dfdlInfoset>
+    </tdml:infoset>
+  </tdml:parserTestCase>
+
+  <!-- xs:dateTime with binaryCalendarRep='binaryMillieconds' -->
+  <tdml:parserTestCase name="dateTimeBin5" root="dateTimeBin5"
```

Comment:
I'm curious now as to what date/time 64-bits of all 1's as binaryMilliseconds is.  I wonder if it can be represented with 4-digits for year. Probably not. I don't know if the Java date/time libraries will work for this. It would be good to have tests that characterize the range that Java's date/time stuff supports. I.e., what is the max 64-bit integer for binary milliseconds that they are able to support, and what kinds of errors occur when the limits are exceeded.

Another interesting dateTime would be 9999-12-31T23-59-59.999999Z  which is the last millisecond of the "4-digit year" era. We should certainly have tests that have that, and the next millisecond after it, which would be
10000-01-01T00-00-00.000000Z.

Reply:
I forgot these are signed. So there are extrema both positive future date/times and past date/times. 

Back of the envelope calculation shows that 64-bits of milliseconds can represent a delta of like 50+ million years, so yeah, they can go way beyond 4-digit years.

Reply:
For very large binaryMilliseconds, I don't see anything in the spec about expected behavior. Right now if adding puts it over the max long it will have a negative date.

Also, a Calendar object can either be in strict or lenient mode. When in strict mode, an IllegalIcuArgumentException  will be thrown for calling setTimeInMillis with a long that is outside of the permitted range. In lenient mode it will give the min or max date permitted (-5838269-09-20 or 5828963-12-20). Default is lenient.

i.e. With binaryCalendarEpoch="2000-06-15T03:25:19" and binaryMilliseconds (as a long) = 9223372036854775807 the resulting date is "-5838269-09-20T00:00:00"

So which do we want:
1. Very large numbers will "loop", using the min and max where necessary.
2. Very large numbers should not loop and should cap at max/min.
3. Make everything strict and throw errors if the date would be higher/lower than the max/min.

Reply:
The following code makes it look like the actual date range is huge:
```
> val format = new java.text.SimpleDateFormat("MMM dd HH:mm:ss yyyy G")
> format.setTimeZone(java.util.TimeZone.getTimeZone("UTC"))
> val cal = java.util.Calendar.getInstance()
> cal.setTimeInMillis(Long.MaxValue)
> format.format(cal.getTime())
Sun Aug 17 02:12:55 EST 292278994 AD
> cal.setTimeInMillis(Long.MinValue)
> format.format(cal.getTime())
Sun Dec 02 11:47:04 EST 292269055 BC
```
So the min/max dates since the unix epoc are nearly 300 million years in the past and future. That's a huge timespan--seems like a very reasonable restriction to say we don't support dates before or after that. I wonder if dates that large even work with our text parsing stuff, I'd guess not.

So I'd vote for option 3, if the number of milliseconds since dfdl:binaryCalendarEpoch is greater or less than Long.MaxLong/MinLong, that should just be an error.

Reply:
For the com.ibm.icu.util.Calendar the range is smaller, but still pretty large. 

"The range of handled dates has been narrowed from +/- ~300,000,000 years to +/- ~5,000,000 years. In practical terms this should not affect clients. However, it does mean that client code cannot be guaranteed well-behaved results with dates such as Date(Long.MIN_VALUE) or Date(Long.MAX_VALUE). Instead, the Calendar protected constants should be used. "

So we would have to error at their max if we go with option 3.

Reply:
5 million is still pretty reasonable. I'd still vote for an error at whatever long value is ICU max.